src/Microsoft.ML.Data/Data/Conversion.cs

```diff
@@ -1389,61 +1326,53 @@ private bool TryParseNonNegative(string text, int ich, int lim, out long result)
 
         /// <summary>
         /// This produces zero for empty. It returns false if the text is not parsable as a signed integer
-        /// or the result overflows. The min legal value is -max. The NA value is -max - 1.
+        /// or the result overflows. The min legal value is -max. The NA value null.
         /// When it returns false, result is set to the NA value. The result can be NA on true return,
         /// since some representations of NA are not considered parse failure.
         /// </summary>
-        private bool TryParseSigned(long max, ref TX span, out long result)
+        private void TryParseSigned(long max, ref TX text, out long? result)
         {
             Contracts.Assert(max > 0);
             Contracts.Assert((max & (max + 1)) == 0);
 
-            if (!span.HasChars)
+            if (text.IsEmpty)
             {
-                if (span.IsNA)
-                    result = -max - 1;
-                else
-                    result = 0;
-                return true;
+                result = default(long);
+                return;
             }
 
-            int ichMin;
-            int ichLim;
-            string text = span.GetRawUnderlyingBufferInfo(out ichMin, out ichLim);
-
-            long val;
+            ulong val;
+            var span = text.Span;
             if (span[0] == '-')
             {
-                if (span.Length == 1 ||
-                    !TryParseNonNegative(text, ichMin + 1, ichLim, out val) ||
-                    val > max)
+                if (span.Length == 1 || !TryParseCore(text.Slice(1, text.Length - 1), out val) || (val > ((ulong)max + 1)))
                 {
-                    result = -max - 1;
-                    return false;
+                    result = null;
+                    return;
                 }
                 Contracts.Assert(val >= 0);
                 result = -(long)val;
-                Contracts.Assert(long.MinValue < result && result <= 0);
-                return true;
+                Contracts.Assert(long.MinValue <= result && result <= 0);
+                return;
             }
 
-            if (!TryParseNonNegative(text, ichMin, ichLim, out val))
+            long sVal;
+            if (!TryParseNonNegative(text.Slice(0), out sVal))
```

Comment:
`text.Slice(0)` is the same thing is `text`, right?  Shouldn't need to `Slice` here.